### ISSUE TITLE
Implement document.all

### ISSUE THREAD

**Issue by frewsxcv** [servo/servo#7396]
https://html.spec.whatwg.org/#dom-document-all

**Comment by frewsxcv** [comment]
I'm working on this

**Comment by frewsxcv** [comment]
Related to https://github.com/servo/servo/issues/3223

**Comment by frewsxcv** [comment]
[The spec](https://html.spec.whatwg.org/multipage/obsolete.html#dom-document-all) makes a mention of "several unusual behaviours" that need to be implemented. @jdm informed me on IRC that there is no infrastructure in-place to support these unusual behaviors, so there's not much point in progressing `document.all` without it. Unassigning.

**Comment by Ms2ger** [comment]
I have no intention of accepting a patch for document.all at this point.

**Comment by metajack** [comment]
@Ms2ger Can you be more specific about why? I'm assuming this is getting deprecated? I saw mention on blink-dev that `if (document.all)` returns falsey, so I'm assuming it's old and crufty.

**Comment by Ms2ger** [comment]
I'm still hoping it can be removed from the platform eventually.

**Comment by nox** [comment]
Even if removed, same behaviour is accessible from passing "*" to GetElementsByTagName().

**Comment by frewsxcv** [comment]
@Ms2ger Is there any public online discussion of this somewhere within WHATWG? It might be nice to have some sort of tracking issue somewhere

**Comment by Ms2ger** [comment]
Dunno. In any case, there are so many more useful things to do right now, that I do not think it is a good use of anyone's time right now, even if we want it eventually.

**Comment by nox** [comment]
https://github.com/whatwg/html/issues/775

Seems like there is consensus now.

**Comment by nox** [comment]
Consensus? Ah ah, no.

**Comment by pshaughn** [comment]
WPT test responsexml-document-properties expects this, as well as the HTMLAllCollection constructor its value is supposed to prototype-match.

**Comment by pshaughn** [comment]
https://html.spec.whatwg.org/dev/common-dom-interfaces.html#the-htmlallcollection-interface possibly the spec has been clarified in the past 3 years? The description here seems coherent (as a description of an interface that is itself not especially coherent).

**Comment by jdm** [comment]
Yes, I'm pretty sure that it's solidified as a part of the web platform.

**Comment by pshaughn** [comment]
This is the most exceptional object in all of Javascript. It works in Firefox, so presumably Spidermonkey has the necessary hooks, but they are hooks used by just this one object and nothing else ever.

**Comment by jdm** [comment]
https://searchfox.org/mozilla-central/source/dom/webidl/HTMLAllCollection.webidl#5
https://searchfox.org/mozilla-central/rev/656899de75c7fdc27c650b701748d508cf6a1bee/dom/bindings/Codegen.py#774-778

**Comment by jdm** [comment]
https://searchfox.org/mozilla-central/rev/656899de75c7fdc27c650b701748d508cf6a1bee/dom/webidl/HTMLAllCollection.webidl#13
https://searchfox.org/mozilla-central/rev/656899de75c7fdc27c650b701748d508cf6a1bee/dom/bindings/Codegen.py#656-660
https://searchfox.org/mozilla-central/rev/656899de75c7fdc27c650b701748d508cf6a1bee/dom/bindings/Codegen.py#10843
https://searchfox.org/mozilla-central/rev/656899de75c7fdc27c650b701748d508cf6a1bee/dom/html/HTMLAllCollection.h#63

**Comment by jdm** [comment]
This should be straightforward to port from Gecko. I'm happy to advise someone who would like to implement it! The new CGLegacyCallHook can use https://github.com/servo/servo/blob/d96b147bab8abacec47c3ce5931f5de8670ca0c0/components/script_bindings/codegen/codegen.py#L6792 as a general model (since we don't have an equivalent of `CGAbstractBindingMethod`). The other linked code from CGDOMProxyJSClass and CGDOMJSClass can be copied directly to the same classes in Servo's codegen.py.
